src/Qir/Runtime/lib/QIR/QubitManager.hpp

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Qubit Manager maintains mapping between user qubit objects and
+// underlying qubit identifiers (Ids). When user program allocates
+// a qubit, Qubit Manager decides whether to allocate a fresh id or
+// reuse existing id that was previously freed. When user program
+// releases a qubit, Qubit Manager tracks it as a free qubit id.
+// Decision to reuse a qubit id is influenced by restricted reuse
+// areas. When a qubit id is freed in one section of a restricted
+// reuse area, it cannot be reused in other section of the same area.
+// Borrowing of qubits is not supported and is currently implemented
+// as plain allocation.
+
+#include <iostream>
+#include <vector>
+#include <limits>
+
+#include "CoreTypes.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+
+    // TODO: Remove this! Use standard ones.
+    static void FailNow(const char* message) { std::cout << message; throw message; }
+    static void FailIf(bool condition, const char* message) {
+        if (condition) {
+            FailNow(message);
+        }
+    }
+
+    using QubitIdType = ::int32_t;
+
+    // The end of free lists are marked with NoneMarker value. It is used like null for pointers.
+    // This value is non-negative just like other values in the free lists.
+    constexpr QubitIdType NoneMarker = std::numeric_limits<QubitIdType>::max();
+
+    // Explicitly allocated qubits are marked with Allocated value.
+    // Qubits implicitly allocated for borrowing are "refcounted" with values [Allocated+1 .. -2].
+    // When refcount needs to be decreased to Allocated value, the qubit is automatically released.
+    constexpr QubitIdType AllocatedMarker = std::numeric_limits<QubitIdType>::max();
```

Comment:
FYI: I'm a bit confused.
* What is "Allocated value"? Did you mean `AllocatedMarker`? An example would help...
* The fragment `[Allocated+1 .. -2]` (the value `-2`) makes an impression, that you meant: 
`[Allocated+1, Allocated+0, Allocated-1, Allocated-2, ... , 2, 1, 0, -1, -2]`
But intuitively I feel that you meant something different.

Reply:
It is AllocatedMarker and it is supposed to be int min value, not max. I also removed the wording about borrowing as it is not supported. We lack some unit tests (which I will add) and some comments about how status array is structured (which I will also add).